pkg/fuzzer/queue/avoid.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package queue
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/syzkaller/pkg/stats"
+)
+
+type Avoider struct {
+	source        Source
+	seq           atomic.Uint64
+	empty         atomic.Bool
+	active        atomic.Pointer[[]atomic.Uint64]
+	mu            sync.Mutex
+	queue         []*Request
+	statDelayed   *stats.Val
+	statUndelayed *stats.Val
+	statViolated  *stats.Val
+}
+
+func Avoid(source Source) *Avoider {
+	return &Avoider{
+		source: source,
+		statDelayed: stats.Create("avoid delayed", "Number of test programs delayed due to VM avoidance",
+			stats.Graph("avoid")),
+		statUndelayed: stats.Create("avoid undelayed", "Number of test programs undelayed for VM avoidance",
+			stats.Graph("avoid")),
+		statViolated: stats.Create("avoid violated", "Number of test programs violated VM avoidance",
+			stats.Graph("avoid")),
+	}
+}
+
+// Next returns the next request to execute on the given vm.
+// VM = -1 allows to disable delaying logic and drain the queue.
+func (avoid *Avoider) Next(vm int) *Request {
+	if vm >= 0 {
+		avoid.noteActive(vm)
+	}
+	if req := avoid.delayed(vm); req != nil {
+		return req
+	}
+	for {
+		req := avoid.source.Next()
+		if req == nil || vm < 0 || !contains(req.Avoid, vm) || !avoid.hasOtherActive(req.Avoid) {
+			return req
+		}
+		avoid.delay(req)
+	}
+}
+
+func (avoid *Avoider) delay(req *Request) {
+	avoid.mu.Lock()
+	defer avoid.mu.Unlock()
+	req.delayedSince = avoid.seq.Load()
+	avoid.queue = append(avoid.queue, req)
+	avoid.statDelayed.Add(1)
+	avoid.empty.Store(false)
+}
+
+func (avoid *Avoider) delayed(vm int) *Request {
+	if avoid.empty.Load() {
+		return nil
+	}
+	avoid.mu.Lock()
+	defer avoid.mu.Unlock()
+	seq := avoid.seq.Load()
+	for i, req := range avoid.queue {
+		violation := contains(req.Avoid, vm)
+		if vm >= 0 && violation && req.delayedSince+1000 > seq {
+			continue
+		}
+		avoid.statUndelayed.Add(1)
+		if violation {
+			avoid.statViolated.Add(1)
+		}
+		last := len(avoid.queue) - 1
+		avoid.queue[i] = avoid.queue[last]
+		avoid.queue[last] = nil
+		avoid.queue = avoid.queue[:last]
+		avoid.empty.Store(len(avoid.queue) == 0)
+		return req
+	}
+	return nil
+}
+
+func (avoid *Avoider) noteActive(vm int) {
```

Comment:
I'm curious if we should be really detecting it here in this queue wrapper. We know the number of fuzzing VMs very well in `syz-manager`, so maybe it would have been easier to enable/disable the functionality explicitly at `RPCServer` construction?

Also, let's consider another case: we had 10 VMs, fuzzed for some time, then started a bug reproduction. Just one fuzzing VM remained, and it will be so for some time. Should we ignore the directions of `deflake()` to ignore the specific VMs? I think it may be wise to not ignore them in this case. Eventually bug reproduction will release some VMs or that single VM will crash/restart, so we will be able to continue the process. This way we ensure that we get high quality corpus items all the time and not just when we have enough fuzzing VMs.

Reply:
> Eventually bug reproduction will release some VMs

Another repro may start right after the previous one, we frequently have a queue of them. As the result they may never be released in this manager run.

> or that single VM will crash/restart

Then it may take up to 5 hours for a single triage, or for emulated qemu up to 50 hours.
If there is just 1 VM, then each inputs triage will take up to 20/200 hours.
This looks as too much time. I would start with using this functionality as best effort if we have active VMs.
We could additionally increase number of always fuzzing VMs.

Reply:
On syzbot, our emulated instances don't really get past the corpus triage stage anyway :) That's not a 100% argument though, I agree.

But we ultimately want to collect and preserve a nice corpus between runs, right? I think it would be nice to be consistent in that during the whole fuzzing time, not just during the times when we have enough fuzzing VMs.

Maybe we can keep track of the triage jobs that are waiting for a different VM and forcefully restart the VM after some threshold?  That should hopefully be not too complicated to implement and the restart times are not particularly high even for the emulated instances.

Reply:
Let's do it separately. (1) we still need all of logic that this change adds (restarts is mostly an addition on top), (2) it's still not obvious to me (smashing inputs may produce new inputs, so there will be a chain of inputs, overall latency can be huge).